lib/Dialect/Arc/Transforms/AllocateState.cpp

```diff
@@ -62,10 +62,27 @@ void AllocateStatePass::allocateBlock(Block *block) {
   }
   LLVM_DEBUG(llvm::dbgs() << "- Visiting block in "
                           << block->getParentOp()->getName() << "\n");
+  
+  // The allocation should run in (reversed?) topo-sorted order for preexisting
+  // substorages to work correctly.
+
+  DenseSet<Value> allocated;
+  std::function<void(Value)> visit;
+  visit = [&](Value parent) {
+    if(allocated.contains(parent)) return;
+    for(auto childAlloc : opsByStorage[parent])
+      if(auto childStorageAlloc = dyn_cast<AllocStorageOp>(childAlloc)) {
+        auto child = childStorageAlloc.getResult();
+        if(opsByStorage.contains(child)) // Used by allocations in this block
+          visit(child);
+      }
+    allocateOps(parent, block, opsByStorage[parent]);
+    allocated.insert(parent);
+  };
 
   // Actually allocate each operation.
-  for (auto &[storage, ops] : opsByStorage)
-    allocateOps(storage, block, ops);
+  for (auto &[storage, _] : opsByStorage)
+    visit(storage);
 }
```

Comment:
Thanks a lot for looking into why this didn't work for your use case! Could you  add a small test case to `test/Dialect/Arc/allocate-state.mlir` that's broken, and create a separate PR together with your fix for the AllocateState pass?

Reply:
We added the test in 7678feb7b, included in this PR. Is it possible that we first got this pr merged into a dev branch, and then break up the changes into smaller PRs?

Thank you!